lib/AST/Module.cpp

```diff
@@ -784,6 +784,56 @@ bool ModuleDecl::isInGeneratedBuffer(SourceLoc loc) {
   return file->Kind == SourceFileKind::MacroExpansion;
 }
 
+std::pair<unsigned, SourceLoc>
+ModuleDecl::getOriginalLocation(SourceLoc loc) const {
+  assert(loc.isValid());
+
+  SourceManager &SM = getASTContext().SourceMgr;
+  unsigned bufferID = SM.findBufferContainingLoc(loc);
+
+  SourceLoc startLoc = loc;
+  unsigned startBufferID = bufferID;
+  while (Optional<GeneratedSourceInfo> info =
+             SM.getGeneratedSourceInfo(bufferID)) {
+    switch (info->kind) {
+    case GeneratedSourceInfo::ExpressionMacroExpansion:
+      LLVM_FALLTHROUGH;
+    case GeneratedSourceInfo::FreestandingDeclMacroExpansion:
+      LLVM_FALLTHROUGH;
+    case GeneratedSourceInfo::AccessorMacroExpansion:
+      LLVM_FALLTHROUGH;
+    case GeneratedSourceInfo::MemberAttributeMacroExpansion:
+      LLVM_FALLTHROUGH;
+    case GeneratedSourceInfo::MemberMacroExpansion:
+      LLVM_FALLTHROUGH;
+    case GeneratedSourceInfo::PeerMacroExpansion:
+      LLVM_FALLTHROUGH;
+    case GeneratedSourceInfo::ConformanceMacroExpansion: {
```

Comment:
Any reason not to write these without the explicit fallthroughs?

```suggestion
    case GeneratedSourceInfo::ExpressionMacroExpansion:
    case GeneratedSourceInfo::FreestandingDeclMacroExpansion:
    case GeneratedSourceInfo::AccessorMacroExpansion:
    case GeneratedSourceInfo::MemberAttributeMacroExpansion:
    case GeneratedSourceInfo::MemberMacroExpansion:
    case GeneratedSourceInfo::PeerMacroExpansion:
    case GeneratedSourceInfo::ConformanceMacroExpansion: {
```

Reply:
Apparently not, I thought I had a warning here but that must have been when I had code in one of them 👍